### PR TITLE
Fix cutting the graph

### PR DESCRIPTION
**Pull request by artkuli**
### Tickets:
 - *84041*

Tested usage:
- `--input "node_name" --output "node_name"`
- `--input "port_number:node_name" --output "node_name"` 
- `--input "node_name:port_number" --output "node_name"` 
- `--input "node_name" --output "port_number:node_name"` 
- `--input "node_name" --output "node_name:port_number"` 
- `--input "node_name"`
- `--input "port_number:node_name"`
- `--input "node_name:port_number"`
- `--output "node_name"`
- `--output "port_number:node_name"`
- `--output "node_name:port_number"`
on model `mnist-8`

### CODE REVIEW

**Review by tomdol — APPROVED** [review]

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py [hunk 1]
```diff
@@ -6,100 +6,105 @@
...
-            if tensor:
-                node_names.append('Tensor:' + tensor.get_names()[0])
-                nodes.append(tensor)
-                search_tensor = False
+    # find by tensor name
+    node = input_model.get_place_by_tensor_name(node_name)
+    if node:
+        found_node_names.append("Tensor:" + node_name)
+        found_nodes.append(node)
+    else:
```
Comment (mlukasze) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:41]: I think that level of cyclomatic complexity is not necessary here.
It could work in a way like this
```
else:
        # find by operation name
        node = input_model.get_place_by_operation_name(node_name)
  
if not node:
    return
	
	
name = node_name
if framework == "onnx" and framework == "onnx":
   if io_type == IOType.Input:
```
Also is worth to note that `node` variable is heavly mutated and it may be hard to understand what contains in each moment (different nodes? different type of nodes?)

**Review by mlukasze — CHANGES_REQUESTED** [review]

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py [hunk 2]
```diff
@@ -6,100 +6,105 @@
...
+                    return tensor
+                return tensor.get_producing_operation()
+        return None
+
+    regexp_post = r"(.+):(\d+)"
+    match_post = re.search(regexp_post, node_name)
+    if match_post:
+        name = match_post.group(1)
+        node_post = try_get_node(input_model, name, framework)
+        if node_post:
```
Comment (mlukasze) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:77]: here similar issue as above - unnecessary nesting

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py [hunk 3]
```diff
@@ -6,100 +6,105 @@
...
+                else:
+                    found_node_names.append(name)
+                    found_nodes.append(output_port)
+
+    regexp_pre = r"(\d+):(.+)"
+    match_pre = re.search(regexp_pre, node_name)
+    if match_pre:
+        name = match_pre.group(2)
+        node_pre = try_get_node(input_model, name, framework)
+        if node_pre:
```
Comment (mlukasze) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:94]: here too - unnecessary nesting

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py [hunk 4]
```diff
@@ -6,100 +6,105 @@
 from openvino.tools.mo.front.extractor import raise_no_node, raise_node_name_collision
 from openvino.tools.mo.utils.error import Error
 
-from openvino.frontend import InputModel  # pylint: disable=no-name-in-module,import-error
+from openvino.frontend import (
```
Comment (mlukasze) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:9]: unnecessary change, provides no logical difference, premature optimization

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py [hunk 5]
```diff
@@ -155,19 +164,36 @@ def fe_input_user_data_repack(input_model: InputModel, input_user_shapes: [None,
...
+                        "node": node,
+                        "shape": shape,
+                        "data_type": data_type,
+                        "input_name": input_name,
+                    }
+                )
             else:
-                _input_shapes.append({'node': node, 'shape': shape, 'input_name': input_name})
+                _input_shapes.append(
+                    {"node": node, "shape": shape, "input_name": input_name}
```
Comment (mlukasze) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:191]: keep same notation - for condition above you split list to new lines, here is half-way changed only

**Review by mlukasze — CHANGES_REQUESTED** [review]

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py [hunk 6]
```diff
@@ -6,100 +6,105 @@
...
-    POST = 2
 
+class IOType(Enum):
+    Input = 1
+    Output = 2
 
-def decode_name_with_port(input_model: InputModel, node_name: str, framework=""):
+
+def decode_name_with_port(
+    input_model: InputModel, node_name: str, framework="", io_type=IOType.Input
```
Comment (p-wysocki) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:24]: You can provide types along with default values, as well as a return type: 
`def foo(opts: dict = {}) -> float:`

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py [hunk 7]
```diff
@@ -109,12 +114,16 @@ def try_get_nodes(input_model, node_name):
...
 
 
-def fe_input_user_data_repack(input_model: InputModel, input_user_shapes: [None, list, dict, np.ndarray],
-                              freeze_placeholder: dict, framework: str, input_user_data_types=dict(), ):
+def fe_input_user_data_repack(
+    input_model: InputModel,
+    input_user_shapes: [None, list, dict, np.ndarray],
+    freeze_placeholder: dict,
+    framework: str,
+    input_user_data_types=dict(),
```
Comment (p-wysocki) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:125]: Mutable default arguments in Python [are not recommended](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/), they can lead to very time-consuming bugs if the variable is being modified in the function body. Even if it is not in this case, it is dangerous for future code maintenance.

Comment (jiwaszki) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py:125]: That is right, it should be resolved at the beginning of the function's body. Use `None` as default argument and later on check for it and replace if necessary.

File: tools/mo/openvino/tools/mo/moc_frontend/pipeline.py [hunk 8]
```diff
@@ -51,9 +51,8 @@ def add_names_to_tensors(model: InputModel, places: List[Place]):
         :param places An object containing Places and names that will be used for model modification
         """
         for new_input in places:
-            if not hasattr(new_input, 'input_name'):
+            if not 'input_name' in new_input:
```
Comment (p-wysocki) [inline tools/mo/openvino/tools/mo/moc_frontend/pipeline.py:54]:
```suggestion
            if 'input_name' not in new_input:
```

Comment (p-wysocki) [inline tools/mo/openvino/tools/mo/moc_frontend/pipeline.py:54]:
```suggestion
            if 'input_name' not in new_input:
```

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py (hunk 5 above)
Comment (artkuli) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py:213]: It is result of using auto formatter `autopep8` on IDE

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py (hunk 4 above)
Comment (artkuli) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py:9]: It is result of using auto formatter `autopep8` on IDE

Comment (artkuli) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py]: Restored to the previous version(before using auto formatter)

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py (hunk 5 above)
Comment (artkuli) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py]: Fixed

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py (hunk 6 above)
Comment (artkuli) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py:23]: Done

File: tools/mo/openvino/tools/mo/moc_frontend/pipeline.py (hunk 8 above)
Comment (artkuli) [reply tools/mo/openvino/tools/mo/moc_frontend/pipeline.py]: Changed

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py (hunk 1 above)
Comment (artkuli) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py:53]: I made a little bit of changes to simplify that part of code but in this case, the early return will not work correctly.
The algorithm of looking for the place has to include that user can pass place name as:
- direct name of operation
- name which contains input/output port of operation, like: <name>:<port> or <port>:<name>
- direct tensor name
So if we don't find node by the direct name of tensor/operation then checking name with port will be not included.
Also algorithm has to check name collision (different places have the same name). If it will be not included,  cutting can work not properly

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py (hunk 7 above)
Comment (artkuli) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py]: Good point, I've not noticed that before, maybe because in this place changes were made by auto formatter. I've changed default argument to `None` , variable `input_user_data_types` is checked in line 205: `isinstance(input_user_data_types, dict)`

**Review by mlukasze — APPROVED** [review]

**Review by p-wysocki — APPROVED** [review]
